### PR TITLE
Change license

### PR DESCRIPTION
**Pull request by patriknw**
https://www.lightbend.com/blog/why-we-are-changing-the-license-for-akka

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -1,212 +1,97 @@
...
+Change License:       Apache License, Version 2.0
+
+For information about alternative licensing arrangements for the Software,
+please visit: https://lightbend.com/
+
+-----------------------------------------------------------------------------
+
+Business Source License 1.1
+
+License text copyright (c) 2017 MariaDB Corporation Ab, All Rights Reserved.
```
Comment (He-Pin) [inline LICENSE:29]: MariaDB not Lightbend?

Comment (patriknw) [reply LICENSE:29]: This copyright is for the license text for "Business Source License 1.1" itself

**Review by johanandren — APPROVED** [review]
LGTM

**Review by octonato — APPROVED** [review]

**Comment by ckipp01** [comment]
Quick question on this. What will the next release tag be that this will apply to? Will this be a patch, minor, etc? I'm sure that many will want to freeze on the last Apache release, so it'll be helpful for companies to know this so they can lock the version using tooling like Scala Steward.

**Comment by johanandren** [comment]
@ckipp01 It will apply to 2.7.0, so next minor (you can see that explicitly expressed in the LICENSE file)

**Comment by He-Pin** [comment]
So sad akka actor and akka stream and akka http is not keeping opensource

**Comment by jboner** [comment]
The last tag under Apache 2 is 2.6.20. 
2.7.x and onward will be under BSL.

**Comment by He-Pin** [comment]
@jboner So the next version is 2.7 not 2.6.21?

**Comment by caraboides** [comment]
@jboner This means  no more security updates for 2.6.x ?

**Comment by He-Pin** [comment]
@jboner Akka is great, would you consider to keep akka-actor akka-stream and akka-http under the Apache L2 license? This will get more people onboard not push away.

**Comment by jlprat** [comment]
I'm concerned about OSS libraries that use Akka underneath and they are Apache v2 licensed. As far as I understand if they decide to upgrade to 2.7 and above they effectively are BSL licensed as well, excluding them to be used in production by commercial companies unless they acquire Lightbend's license. Is my assumption right, or am I missing something?
Sorry to hijack the PR for this question.

**Comment by mdedetrich** [comment]
> I'm concerned about OSS libraries that use Akka underneath and they are Apache v2 licensed. As far as I understand if they decide to upgrade to 2.7 and above they effectively are BSL licensed as well, excluding them to be used in production by commercial companies unless they acquire Lightbend's license. Is my assumption right, or am I missing something?
Sorry to hijack the PR for this question.

To add ontop of this, if an exemption is granted is it transitive? i.e. If some OS library that uses akka happens to get an exemption like Play has, does that mean that any other project that uses that OS library also has the same exemption even if they use it commercially in production?

Or does it mean regardless if some OS library that uses akka in your stack has an exemption (or not), you simply put have to pay for it for commercial use (i.e. if I just happen to have akka in my classpath I am already forced to pay for it).

**Comment by jboner** [comment]
> @jboner This means no more security updates for 2.6.x ?

Correct. As stated in the [FAQ](https://www.lightbend.com/akka/license-faq)
Q: Can I continue to use the version of Akka I’m currently using?
A: Yes. However there will be no additional features, enhancements, security updates, or bug fixes for existing versions."

**Comment by jboner** [comment]
> @jboner So the next version is 2.7 not 2.6.21?

Yes.

**Comment by kyri-petrou** [comment]
My reaction to the news about the licensing change

redshirtdude

**Comment by domdorn** [comment]
@jboner you realize that not providing at least security updates for 2.6.x basically leaves the community with no other choice than to fork, right? or will there be a 2.6.x branch which accepts further community contributions, e.g. allowing contributors to fix security issues by themselves? this is a very concerning move of lightbend.

**Comment by nate100guess** [comment]
Help me

**Comment by Khalian** [comment]
If Lightbend refuses to even entertain 0 day exploits/CVEs in older versions entirely or try to get in the way in back porting those fixes, then Lightbend is basically an untrustworthy company and I will now have to spend time trying to de platform Akka out of our production environments entirely.

**Comment by 1zizi1** [comment]
Congratulations on this brave decision!

Those who disliked this are the same people who cherish colonization and the exploitation of free labour.

**Comment by He-Pin** [comment]
@nate100guess Send a PR , help the team

**Comment by jboner** [comment]
@domdorn See this post and the decision regarding addressing critical security updates and critical bugs in Akka v2.6.x: https://discuss.lightbend.com/t/2-6-x-maintenance-proposal/9949/2

**Comment by domdorn** [comment]
Thanks @jboner! It would be good to spread the word about this to minimize damage to akka and the whole scala ecosystem - I'll do my part!

**Comment by jboner** [comment]
@domdorn Thanks

**Comment by yasha-spare** [comment]
The pricing for this is insane - a minimum of $2K/year per core/vCore/vCPU? That means if you're running apps that use Akka across, say, a not-that-big K8s cluster with ~6 machines, each ~8 cores, you'd be paying ~$100,000/year just to use a library. That is a ludicrous price for using a library.

IMO this is pure Oracle-esque rent-seeking behaviour, but at least people are willing to pay huge prices for DBs, they aren't for libraries. Feels like you're just trying to get a short-term revenue boost from companies that adopted Akka as a true open source library, and are now too invested to move off of it. Going forward, I certainly would avoid any Lightbend library like the plague, even for <$25 mil annual revenue companies, as you've shown that you'll simply change terms to squeeze $$. You're probably killing longer term adoption of all Lightbend software just to pump up short term revenue numbers.

**Comment by yashap** [comment]
The pricing for this is insane - [a minimum of $2K/year per core/vCore/vCPU](https://www.lightbend.com/akka#pricing)? That means if you're running apps that use Akka across, say, a not-that-big K8s cluster with **_~6 machines, each ~8 cores, you'd be paying ~$100,000/year just to use a library_**. That is a ludicrous price.

IMO this is pure Oracle-esque rent-seeking behaviour, but at least people are willing to pay huge prices for DBs, they aren't for libraries. Feels like you're just trying to get a short-term revenue boost from companies that adopted Akka as a true open source library, and are now too invested to move off of it. Going forward, I certainly would avoid any Lightbend library like the plague, even for <$25 mil annual revenue companies, as you've shown that you'll simply change terms to squeeze $$. You're probably killing longer term adoption of all Lightbend software just to pump up short term revenue numbers.

Especially sleazy given non-Lightbend ppl have contributed to Akka assuming it's a true open source project, as it used to be.

**Comment by Meligy** [comment]
I can't be the only one seeing this but such licence change is certainly a breaking change for those relying on the old licence. Hence, it should be a major version bump.

**Comment by PerWiklander** [comment]
Yeah, bumping the version with the new license to `3.0.0` seems logical to make it clear that this is a whole new product.

It would also be fair to contributors to merge and release all PRs (that would be accepted) submitted before the news of the new license under the old, Apache 2 license. [He-Pin](https://github.com/akka/akka/issues?q=is%3Apr+is%3Aopen+author%3AHe-Pin) submitted a batch of PRs recently. Part of those were released but part of them will be stuck under the new license and I suspect they won't even be allowed to use their own contributed code without paying for it now.

**Comment by mdedetrich** [comment]
> It would also be fair to contributors to merge and release all PRs (that would be accepted) submitted before the news of the new license under the old, Apache 2 license. [He-Pin](https://github.com/akka/akka/issues?q=is%3Apr+is%3Aopen+author%3AHe-Pin) submitted a batch of PRs recently. Part of those were released but part of them will be stuck under the new license and I suspect they won't even be allowed to use their own contributed code without paying for it now.

I am in the same boat, had open PRs for Akka but I closed them because I cannot use them with the new license due to legal issues. If Lightbend is willing to merge those PRs into the 2.6 Apachr branch I can reopen them.

**Comment by He-Pin** [comment]
@patriknw @PerWiklander @jboner @mdedetrich 
I think it would be fair that:
1.  Cut out a new branch `release-2.6` which keeps `Apache 2` license.
2. All contributors can still make contributions to the `release-2.6` and `main` branch.
3. Lightbend core team may take sometime to merge these PR and not let these PRs wasted and the time we the community spend on it.
4. As the CLA is been signed, so its totally OK cherry pick all the merged PRs from the community to the current `main`, the next `2.7` release.
5. New Features from Lightbend core team only went into the `main` branch.
6. After one year, the community can fork(???), contributes both branch too? Who knows.

I think these would be more fair for all in the community. There are PRs closed just because that, please.

**Comment by mdedetrich** [comment]
Thats perfectly reasonable for me and with a new `release-2.6` branch its easy for the community to just target the PR's onto that branch

**Comment by He-Pin** [comment]
For me too, there are PRs comes from China Scala User Group and I do contributes my weekend to this projects and advise nearly others (500+) members too.

And Yes I can hardly let the every company I am working at gives money, at least that what I can do.

**Comment by PerWiklander** [comment]
> The worst case of this is [akka/akka-http#4079](https://github.com/akka/akka-http/pull/4079). Realistically right now that entire effort is wasted (I doubt the contributors will be willing to put the changes into the new BSL license)

Ugh, that one would be really bad to loose. Arguably it would be even worse if it is merged now after the split. Would it even be possible to contribute Scala 3 support for the community if that one is merged into 2.7? It will be quite hard to unsee the solution and do a clean room implementation that can't be said to be stolen from the BSL version.

Us waiting until 2027 (?) to transition projects to Scala 3 is probably not what Odersky had in mind.

**Comment by mdedetrich** [comment]
> would it even be possible to contribute Scala 3 support for the community if that one is merged into 2.7? It will be quite hard to unsee the solution and do a clean room implementation that can't be said to be stolen from the BSL version.

Afaik (note IANAL) it would be a legal minefield. Basically anyone that worked on implementing Scala 3 in the Apache version of akka-http would have to prove their code is not derived/copied from the BSL version of akka-http (which practically means proving that the person has never read/written the BSL code).

**Comment by blitzcaster** [comment]
Wow, I have a very early-stage hobby project using Akka under AGPL v3 license. I'm considering myself very lucky since it's still not going anywhere. Still quite a shame that I have to scrap it completely now due to incompatible licensing.

**Comment by spangaer** [comment]
> It will be quite hard to unsee the solution and do a clean room implementation that can't be said to be stolen from the BSL version.

> Basically anyone that worked on implementing Scala 3 in the Apache version of akka-http would have to prove their code is not derived/copied from the BSL version of akka-http

Now I didn't study the PR, but afaik the CLA only sub-licenses the code to Lightbend, so any code not written by non-Lightbend employees can just be contributed by the original author under his own copyright.

That being said. The BSL remains a toxic element. When the Sept 2023 commitment expires an open source fork which implements an independent security fix will always be exposed to the same challenge: proving that it's not a BSL derivative. Which is close to impossible. It seems to me that Lightbend should only make it's source available to it's paying customers at that point, if any.

It's also painful to read a blog-post that talks about contributing as a requirement for a working community, to justify an action that hits those that did contribute and have invested a lot in this the hardest.

**Comment by spangaer** [comment]
> so any code not written by non-Lightbend employees can just be contributed by the original author under his own copyright.

Come to think of it, any commit on the PR, which doesn't have the BSL license yet, qualifies as "a copy" of the source to which the ASLv2 still applies. So no unseeing needed, just copying form the right place. It's copyright after all 🙃 .

**Comment by He-Pin** [comment]
I think anycode comes from the community or still licensed as Apache 2 should be free to read and share.

**Comment by He-Pin** [comment]
So opensourcing is copy~~

**Comment by domdorn** [comment]
as for akka-http scala3: 
just change the merge target for that PR from master/main to the 2.6.x/10.2 branch.. this can be released under 2.6.x apache and lightbend can merge/cherry pick to main/master and have it in the 2.7.x/3.x/10.3 branch.. WIN-WIN.. anyway, (IMHO) this should be discussed in the akka-http PR and not in this one. 

As for the general discussion in this thread: IMO the most important short- to middle-term issue (security of user-facing stuff, e.g. akka-http, akka-grpc, etc. ) has been resolved. I'm really thankful for @jboner for being so forthcoming!

If I'd be to start a startup now, I'd even go with the BSL and akka 2.7.x.. if I base my company on akka and I ever hit the 25mil revenue, I'll gladly pay lightbend a share.. most startups however will never come close to this number... 
Unfortunately - for my current customer who's not a startup and I'm "just a dev" -  I cannot justify such an expense for the "http client/server lib" I introduced 2 years ago... would be different if we'd use the whole ecosystem, which we're not.

**Comment by mr-git** [comment]
Pardon my skepticism, I might have missed details in discussion(s) and/or blog(s), but is there a promise from Lightbend that Lightbend will accept, merge and release PRs targeted for akka 2.6.x based libraries?

My observation with other open-source projects is that there are gatekeepers, who can and do veto some PRs.

**Comment by PerWiklander** [comment]
> Pardon my skepticism, I might have missed details in discussion(s) and/or blog(s), but is there a promise from Lightbend that Lightbend will accept, merge and release PRs targeted for akka 2.6.x based libraries?

No. There is a commitment from Lightbend to supply critical security and bug fixes during a period of one year from now. Nothing about what you mention.

**Comment by spangaer** [comment]
> critical security updates and critical bugs

Seems to read closer to what @mr-git says TBH.

**Comment by PerWiklander** [comment]
Of course said bug fixes could have been contributed through PRs by outside contributors. But I don't think that should be mixed up with the notion of Lightbend accepting feature PRs targeting anything  < 2.7.

Here is the news about the maintenance commitment
https://discuss.lightbend.com/t/2-6-x-maintenance-proposal/9949/2

**Comment by Meligy** [comment]
Sorry to be a pain, but it seems very obvious...
New non-compatible license = New Major version

A minor version should only include non-breaking additional features and fixes.

**Comment by nvollmar** [comment]
I can understand the position of Lightbend, but this change makes it very hard to argue for time to make contributions going forward.

**Comment by Khalian** [comment]
I have some questions for Lightbend devs/managers about the practicality of this pricing plan, particularly in auto scaling environments. @jboner If you do not mind? Or if you mind I can move this question to a Lightbend discussion as well if this forum is not appropriate. 

Lets say I have a fleet of X instances each with N cores at steady state running on an auto scalable fleet. Each of those instances run some software that uses Akka as a direct or transitive dependency. 

1. Lets say I have an auto scaling event and increase by Y instances from X due to some unpredictable workload. Would the pricing be for (X + Y) * N cores? Even if the auto scaling event was a temporary spike. 

2. Lets say I have over committed my fleet and wish to remove Z instances from the original X. Would the pricing be ? X * N or (X - Z) * N? I would presume the pricing would still be for the higher of the two but the current language is not clear. Plus is the pricing on the ceiling on all capacity utilization for a calendar year? 

3. How would light bend enforce the pricing in a variable workload environment? There are two obvious problems here
    3.1. A corporation/client of Lightbend could very well try to use Akka truthfully and still get their estimates on how many cores they use wrong, and they would most likely get it wrong entirely for a variable workload system. 
    3.2. There is the obvious risk that a corporation will try to under commit their core usage to Lightbend as well (as in Lightbend cant and should not assume good intentions here) and there is no way for Lightbend to refute or assert a claim if say corporation runs something like local akka actors in a proprietary code base. 

To me the pricing model is ambiguous right now and there are tons of unanswered questions that needs to be clarified in very explicit terms. 

Thanks.

**Comment by mr-git** [comment]
>How would light bend enforce the pricing in a variable workload environment? 
I have an idea: there will be a daemon running in akka, which will phone home with actual number of available CPU cores every minute or so and Lightbend will send a bill for max reported number during the year :) 

With rolling updates it will be even more fun - instead of regular X instances we will run X+<maxSurge> instances!

**Comment by Khalian** [comment]
> > How would light bend enforce the pricing in a variable workload environment?
> 
> I have an idea: there will be a daemon running in akka, which will phone home with actual number of available CPU cores every minute or so and Lightbend will send a bill for max reported number during the year :)
> 
> With rolling updates it will be even more fun - instead of regular X instances we will run X+<maxSurge> instances!

So that will work but not in all cases.

1. It definitely wont work in an isolated subnet in which there are firewall rules to block egress traffic out of ports that this daemon can potentially use. And if this daemon becomes mandatory, would that mean that all use cases of Akka inside isolated subnets just outright not be viable? 

2. Will Akka actors/streams/http etc throw up an exception for all cases in which this daemon stops for whatever reasons? What about network partitions? If Akka mandates that the daemon must be able to send messages to some endpoint lightbend owns, then it would be an availability risk to all consumers. And if the daemon just keeps silently failing in an edge case and the library still runs, could someone just put a firewall rule to block egress? 

I am not so sold on Akka local actors for instance introducing a tacit side effect to be a good architectural choice. 

This daemon would also have to account for multiple environments, for instance different type 1 and 2 hypervisors (Firecracker, QEMU, XEN etc) etc not to mention container environments like docker (on top of just bare metal executions). This isn't an issue per say, but it would be significantly costly and complex to maintain to report the number of CPU cores as platforms evolve. CPU core computation in a shared vCPU env isnt trivial.

**Comment by He-Pin** [comment]
it's that simple true, @Khalian are you currently using Akka , do you have a subscription?

**Comment by mpetricek-corp** [comment]
> What about someone using Akka in Android application？

Unless they qualify for one of the few license exemptions (company small enough, approved usage through play internals, etc ...), the price would be $1995 per core. With typical better smartphone having 4-8 cores, you'll have to pay between $7980 and $15960 **yearly** for each user of your application. IMHO a bit steep price for the value.

**Comment by PerWiklander** [comment]
> How would light bend enforce the pricing in a variable workload environment?
> I have an idea: there will be a daemon running in akka, which will phone home with actual number of available CPU cores every minute or so and Lightbend will send a bill for max reported number during the year :)

I really, really hope that "idea" was a joke :-)

Introducing something like that would ensure I don't use Akka, even if I wanted to pay for it.

**Comment by mr-git** [comment]
I have other idea - we can use shell companies!
*

**Comment by spangaer** [comment]
I wouldn't worry too much about phone home daemons. This is about reeling for big fish, best effort approximations and negotiations. Which is good enough because the ones you're targeting are big corporations which worry that the cost of litigation will be even bigger than just paying up. And even if the estimation is off, it's still a customer and still big money. So, you just agree to do better next time. If there's no ill will, it all just works.

But that’s all based on the theory that an Akka project is some kind of cash cow at a multi-million dollar company. If Akka and Akka HTTP just ended up being an alternative to Tomcat, Jetty or Spring. Trying something new for HTTP/IO piping in a cloud project, which is tiny on the company scale. In that case the financial value is all but spectacular.

In our case enough appreciated that we contributed where that happened to fit. Money was never exchanged because whatever made sense for us, never met the threshold of interest for Lightbend. (We did try btw)

By mistaking Akka for a community project, overnight that belief has become a liability and ransom note for our own work. Half a decade for two of us, a couple more since then. As it turns out now, a very painful misunderstanding on our part.

I fear that, without a community Akka that is shielded from this BS-license, no other outcome then stripping it all out is viable. If so, I’m not even sure Scala will survive this at my employer. If there’s more like us, I doubt it will do the Scala community any good.

> I don't think that too many people use 'enterprise-level' features of Akka like clustering, sharding, persistence, CQRS, projections etc.

Akka Persistence was one of the things that got us on board at the time. I'd like to agree on the other stuff, yet that is what Lightbend did try before. Akka cluster used to be paid for and then it was Lagom and their monitoring suite. Hard to tell them to do the model that didn't work out...

**Comment by makkarpov** [comment]
>I fear that, without a community Akka that is shielded from this BS-license, no other outcome then stripping it all out is viable. If so, I’m not even sure Scala will survive this at my employer. If there’s more like us, I doubt it will do the Scala community any good.

This bears a bit of similarity with Ruby on Rails. In some imaginary world where Rails switched to BS-license Ruby will disappear. I mean, fully cease to exist - no one uses Ruby without Rails. I fear that even Pascal could easily become more popular than Ruby.

Scala is not so tightly coupled to Akka, but Akka still plays very essential role in answering questions like "Why should I choose Scala?". I was asked some time ago what is better than Spring. Turns out that saying "Scala + akka-http" was a mistake, and Scala is never going to be adopted there. I was also asked "how could we do event sourcing better than putting random events into Postgres?". Turns out that recommending akka-persistence was a mistake too, and Scala probably isn't going there also.

**Comment by Khalian** [comment]
> it's that simple true, @Khalian are you currently using Akka , do you have a subscription? What about someone using Akka in Android application？

I am using akka but i do not have a lightbend subscription. I use it just like I would use any other open source library. Android apps have an even surface area of an issue so I included that in the lightbend discuss copy of my original question https://discuss.lightbend.com/t/questions-about-lightbends-practicality-in-the-new-license-pricing-in-variable-workload-autoscaling-dynamic-fleet-environments/9976

**Comment by jboner** [comment]
Hi everyone. 
Thanks for raising your questions. 

With most enterprise customers using Lightbend software, we expect that many particular circumstances can arise. There would be no way to reasonably post all the combinations of the usage needs that our customers have, and the posted pricing is a baseline for very simple deployment. A highly elastic system or distributing instances to end-users are great examples of when the baseline pricing doesn't quite work. In elastic environments, we know usage can fluctuate a lot, which is a concept we encourage. In end-user applications, you may have no way of knowing how many cores your user may have. 

In the cases you have presented, we would encourage you to reach out to us with your needs, and we can customize a package that makes sense and is reasonable. We certainly don't want to be sticklers for the exact core count or introduce complex formulas and tracking for usage. 

We have a great sales team that can work through these types of questions with you. You can contact us via this link: [https://www.lightbend.com/akka#form-section](https://www.lightbend.com/akka#form-section)

**Comment by Khalian** [comment]
@jboner 

So I am genuinely confused right now. The pricing specifically said the pricing would be per core but then there isn't any concrete/reified definition of what that means for the general case (or really for any special cases either)? Can we get some documentation on how the pricing is intended to work for a given calendar year without having to work this out on a case by case basis for each corporation? 

A corporation can have multi dimensional use cases (e.g. one team within a corporation could use it in Android apps and one team could use it to run reactive streams of messages for processing and another could run it for synchronization needs). 

How would a simple form here work out then? Is the expectation here that large enterprise customers aggregate all their disparate use case for one unified system (note that each use case could have a distinct pattern of usage, distinct economics behind it and distinct mechanism by which they depend on Akka, either directly or transitively). 

I guess I do not understand how we can unify a single pricing plan for a heterogenous system at scale. Or is the intent not even supposed to have a unified pricing plan?

**Comment by mahadi-hossain** [comment]
keep them away from new license.

akka-actor
akka-actor-typed
akka-stream
akka-http

Nobody will love to work with playframework without these. And if we have to use playframework with concern of facing problem in future, it will be best decision to avoid Scala for any projects. It proves Scala is going to a dead language very soon.

**Comment by mehmetsalgar** [comment]
> keep these away from new license.
> 
> akka-actor akka-actor-typed akka-stream akka-http
> 
> Nobody will love to work with playframework without these. And if we have to use playframework with concern of facing problem in future, it will be best decision to avoid Scala for any projects. It proves Scala is going to a dead language very soon.

I am afraid you might be right, would Scala die, I don't know, it is equally important for Machine Learning frameworks so it will definitely survive there.

But the biggest selling point of Akka is gone, Enterprises are in critical point, their monoliths that are running for the last 20 years, need adaptations/migrations to efficiently go to cloud. I am involved with several such projects, it is always a fight of do a shift up old J2EE applications with huge Relational Databases to cloud with brute force or do something sensible with modern concepts and frameworks like Akka.

Such migration to Akka, Akka being Open Source was the biggest selling point, you don't have to pay the license costs of commercial J2EE platforms. Rewriting/Modernisation an application new is always a risk, they might want to take that risk for Open Source platform but if they have to pay and take the risks they will probably go for brute force method.

So for me, convincing people to use Akka become much harder :(

**Pull request merged**